odl/util/numerics.py

```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+            n_remove = n_orig - n_new
+            istart = num_l
+            istop = n_orig - (n_remove - istart)
+            arr_slc.append(slice(istart, istop))
+            out_slc.append(slice(None))
+
+            # Make trivial entries for padding slice lists
+            pad_l_slc.append(slice(0))
+            pad_r_slc.append(slice(0))
+
+        elif n_new > n_orig:
+            # Padding case: calculate start and stop indices for the
+            # bigger new array, together with the padding slices
+            n_add = n_new - n_orig
+            istart = num_l
+            istop = n_new - (n_add - istart)
+
+            n_pad_l = len(range(istart))
+            n_pad_r = len(range(istop, n_new))
+
+            # Handle some error scenarios with illegal lengths
+            if pad_mode == 'order0' and n_orig == 0:
+                raise ValueError('in axis {}: need at least 1 value for '
+                                 'order 0 padding, got 0'
+                                 ''.format(i))
+
+            if pad_mode == 'order1' and n_orig == 1:
+                raise ValueError('in axis {}: need at least 2 values for '
+                                 'order 1 padding, got 1'
+                                 ''.format(i))
+
+            for lr, pad_len in [('left', n_pad_l), ('right', n_pad_r)]:
+                if pad_mode == 'periodic' and pad_len > n_orig:
+                    raise ValueError('in axis {}: {} padding length {} '
+                                     'exceeds original array size {}; this is '
+                                     'not allowed for periodic padding'
+                                     ''.format(i, lr, pad_len, n_orig))
+
+                elif pad_mode == 'symmetric' and pad_len >= n_orig:
+                    raise ValueError('in axis {}: {} padding length {} '
+                                     'larger or equal to the original array '
+                                     'size {}; this is not allowed for '
+                                     'symmetric padding'
+                                     ''.format(i, lr, pad_len, n_orig))
+
+            arr_slc.append(slice(None))
+            out_slc.append(slice(istart, istop))
+            pad_l_slc.append(slice(istart))
+            pad_r_slc.append(slice(istop, None))
+
+        else:
+            arr_slc.append(slice(None))
+            out_slc.append(slice(None))
+            pad_l_slc.append(slice(0))
+            pad_r_slc.append(slice(0))
+
+    # Set the "inner" part
+    out[tuple(out_slc)] = arr[tuple(arr_slc)]
+
+    # Perform the padding
+    for i, (slc_l, slc_r) in enumerate(zip(pad_l_slc, pad_r_slc)):
+        if direction == 'forward':
+            _apply_padding(out, slc_l, slc_r, i, pad_mode, pad_const)
+        else:  # direction == 'adjoint'
+            _apply_adjoint_padding(arr, out, slc_l, slc_r, i, pad_mode)
+
+    return out
+
+
+def _slice_dict(axis, ax_len, ndim, slc_l, slc_r):
```

Comment:
Very weird way of making a function. Why not simply make it take the padding variant as an argument and return the actual slice?